src/app/modules/payments/__tests__/stripe.service.spec.ts

```diff
@@ -0,0 +1,563 @@
+    object: 'event',
+    created: 1677205973,
+    data: {
+      object: {
+        id: 'po_MOCK_PAYOUT',
+        object: 'payout',
+        amount: 12345,
+        created: 1677205972,
+        status: 'pending',
+        arrival_date: 1677215972,
+      },
+    },
+    type: 'payout.created',
+  },
+  {
+    id: 'evt_PAYOUT_CANCELLED',
+    object: 'event',
+    created: 1677205983,
+    data: {
+      object: {
+        id: 'po_MOCK_PAYOUT',
+        object: 'payout',
+        amount: 12345,
+        created: 1677205972,
+        status: 'canceled',
+        arrival_date: 1677215972,
+      },
+    },
+    type: 'payout.canceled',
+  },
+] as unknown as Stripe.Event[]
+
+const MOCK_STRIPE_EVENTS_MAP = keyBy(MOCK_STRIPE_EVENTS, 'id')
+
+describe('stripe.service', () => {
+  beforeAll(async () => await dbHandler.connect())
+  afterAll(async () => await dbHandler.closeDatabase())
+  beforeEach(() => jest.clearAllMocks())
+
+  describe('processStripeEvent', () => {
+    let payment: IPaymentSchema
+
+    beforeEach(async () => {
+      await dbHandler.clearCollection(Payment.collection.name)
+      await dbHandler.clearCollection(EncryptPendingSubmission.collection.name)
+      await dbHandler.clearCollection(Submission.collection.name)
+
+      // Arrange
+      // Create a new pending submission and payment
+      const pendingSubmission = await EncryptPendingSubmission.create({
+        submissionType: SubmissionType.Encrypt,
+        form: MOCK_FORM_ID,
+        encryptedContent: 'some random encrypted content',
+        version: 1,
+        responseHash: 'hash',
+        responseSalt: 'salt',
+      })
+      payment = await Payment.create({
+        target_account_id: 'acct_MOCK_ACCOUNT_ID',
+        pendingSubmissionId: pendingSubmission._id,
+        amount: 12345,
+        status: PaymentStatus.Pending,
+        paymentIntentId: 'pi_MOCK_PAYMENT_INTENT',
+        email: 'formsg@tech.gov.sg',
+      })
+      await pendingSubmission.updateOne({
+        paymentId: payment._id,
+      })
+    })
+    afterEach(() => jest.clearAllMocks())
+
+    it('should update the charge status from Pending to Failed when a charge.failed event is received', async () => {
+      // Arrange
+      // Inject the expected webhook logs and state into the payment object.
+      await Payment.updateOne(
+        { _id: payment._id },
+        {
+          webhookLog: [MOCK_STRIPE_EVENTS_MAP['evt_PAYMENT_INTENT_CREATED']],
+        },
+      ).exec()
+
+      // Act
+      const eventChargeFailed = MOCK_STRIPE_EVENTS_MAP['evt_CHARGE_FAILED']
+      const chargeFailed = eventChargeFailed.data.object as Stripe.Charge
+      const result = await StripeService.processStripeEvent(
+        String(payment._id),
+        eventChargeFailed,
+      )
+
+      // Assert
+      expect(result.isOk()).toEqual(true)
+
+      const updatedPayment = await Payment.findById(payment.id)
+      if (!updatedPayment) throw new Error('Payment not found!')
```

Comment:
Is throwing errors from tests expected (pun intended)? If we are asserting that `updatedPayment` exist, it might work if we use `expect(updatedPayment).toBeTruthy()`

Reply:
the intention of throwing the error was to distinguish between a functional failure and a test case error. In this case, it should be impossible for the payment to not exist, and not really something we are expecting from the test case... UNLESS the tests should also verify that the payment document was not deleted by the function, but that seemed a bit extreme haha. tbh i'm not entirely sure what the right thing to do here is, what do you think?

Reply:
> the tests should also verify that the payment document was not deleted by the function,

I feel that it is reasonable to verify it, if we want to ensure that `await Payment.findById(payment.id)` will always return something.

Reply:
[1ffc37a](https://github.com/opengovsg/FormSG/pull/5937/commits/1ffc37ae1dc884e4fc94dd737e0098c2a097fc46)